include/FockSpace/EvaluationIterator.hpp

```diff
@@ -0,0 +1,295 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+// 
+// GQCG-gqcp is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with GQCG-gqcp.  If not, see <http://www.gnu.org/licenses/>.
+// 
+#pragma once
+
+
+#include "Mathematical/Representation/SquareMatrix.hpp"
+
+#include <Eigen/Sparse>
+
+
+namespace GQCP {
+
+
+/**
+ *  A templated private class for Fock spaces (private constructors with Fock spaces friends)
+ *  It supports dense, sparse and matrix vector product storage for elements evaluated in the Fock space.
+ *
+ *  @tparam Matrix              the type of matrix in which the evaluations of the Fock space will be stored
+ */
+template<class Matrix>
+class EvaluationIterator {
+    size_t index = 0;  // current position of the iterator in the dimension of the Fock space
+    size_t end;  // total dimension
+
+    Matrix matrix;  // matrix containing the evaluations
+
+    // CONSTRUCTOR
+    /**
+     * @param dimension         the dimensions of the matrix (equal to that of the fock space)
+     */
+    EvaluationIterator(size_t dimension) :  matrix(Matrix::Zero(dimension, dimension)), end(dimension) {}
+
+
+    // PUBLIC METHODS
+    /**
+     *  Adds a value in which the set index corresponds to the row and a given index corresponds to the column
+     * 
+     *  @param j         column index of the matrix
+     *  @param value     the value which is added to a given position in the matrix
+     */
+    void add_columnwise(size_t j, double value) {
+        this->matrix(this->index, j) += value;
+    }
+
+    /**
+     *  Adds a value in which the set index corresponds to the column and a given index corresponds to the row
+     * 
+     *  @param i         row index of the matrix
+     *  @param value     the value which is added to a given position in the matrix
+     */
+    void add_rowwise(size_t i, double value) {
+        this->matrix(i, this->index) += value;
+    }
+
+    /**
+     *  Move to the next index in the iteration
+     */
+    void increment() {
+        this->index++;
+    }
+
+    /**
+     *  Tests if the iteration is finished, if true the index is reset to 0 
+     * 
+     *  @return true if the iteration is finished
+     */
+    bool is_finished() {
+        if (index == end) {
+            this->index = 0;
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    // GETTER
+    const Matrix& evaluation() const { return this->matrix; }
+
+    // Friend Classes
+    friend class FockSpace;
+    friend class SelectedFockSpace;
+    friend class ProductFockSpace;
+};
+
+
+
+/**
```

Comment:
Update the following classes also in the same line of thought as the previous comments.